grpc/server_test.go

```diff
@@ -25,19 +26,21 @@ func Test_FetchAuthDataSync(t *testing.T) {
 					"endpoint_1": {
 						EndpointId: "endpoint_1",
 						Auth: &proto.Auth{
-							AuthType: proto.Auth_API_KEY_AUTH,
-							AuthTypeDetails: &proto.Auth_ApiKey{
-								ApiKey: "secret_key_1",
+							AuthType: proto.Auth_AUTH_TYPE_API_KEY,
```

Comment:
Isn't this field redundant? There could be an `AuthDetails` field defined in the corresponding `.proto` file, with pointers to 3 different structs: e.g. `AuthTypeAPIKey`, `AuthTypeJWT`, etc.
Then the non-nil pointer in `AuthDetails` would indicate the type of authentication that is being used.

Reply:
I don't believe that it is redundant, as there are multiple places in the code where we need to know the type of Authorization the endpoint is using and having a single field that indicates the type simplifies the code in those places.

In addition, this field allows us to perform schema validation against the YAML structure, eg. API key types may not have JWT-specified fields and vice versa. 

For these reasons I recommend we keep this field.

Reply:
tl;dr Reading through the thread, the way I see it is that both of you are right!

@commoddity Lmk what you think of the compromise below. We can have it as an UPNEXT to avoid slowing down this PR.

---

> In addition, this field allows us to perform schema validation against the YAML structure

@commoddity  This is great, but I'd rather we focus on the core benefits rather than side benefits.

> I don't believe that it is redundant, as there are multiple places in the code where we need to know the type of Authorization the endpoint is using and having a single field that indicates the type simplifies the code in those places.

@commoddity This is possible by leveraging the type of the `oneof` field as well. If it's not something you've done in the past, I can put together a small prototype.

EDIT: I got ChatGPT to do it: https://chatgpt.com/share/674e7811-c164-8002-b6b2-cc38d39b99ca

> [...] pointers to 3 different structs: e.g. AuthTypeAPIKey, AuthTypeJWT, etc. [...]

@adshmh I see where you're coming from, but the `oneof` design we have in place right now is the way to go IMO. I've copy-pasted the proto from the other PR for reference below

----

To what @adshmh was saying, have multiple source of truth (i.e. redundant fields) can lead to downstream complexity and cognitive overhead on users.

For example, I could set conflicting `auth_type` and `auth_type_details`.

I realize this can be validated through code, but I'd rather delete code than add code where possible.

---

This is what we have now:

```proto
message Auth {
  enum AuthType {
    AUTH_TYPE_UNSPECIFIED = 0;
    AUTH_TYPE_API_KEY = 1;
    AUTH_TYPE_JWT = 2;
  }

  // The type of authorization being used.
  AuthType auth_type = 1;

  oneof auth_type_details {
    // Indicates that the GatewayEndpoint does not require authorization.
    NoAuth no_auth = 2;
    // Indicates that the GatewayEndpoint uses a static API key for authorization.
    StaticAPIKey static_api_key = 3;
    // Indicates that the GatewayEndpoint uses an auth provider issued JWT for authorization.
    JWT jwt = 4;
  }
}
```

I believe it can be simplified to this:

```proto
message Auth {
  oneof auth_type_details {
    // Indicates that the GatewayEndpoint does not require authorization.
    NoAuth no_auth = 1;
    // Indicates that the GatewayEndpoint uses a static API key for authorization.
    StaticAPIKey static_api_key = 2;
    // Indicates that the GatewayEndpoint uses an auth provider issued JWT for authorization.
    JWT jwt = 3;
  }
}
```

And use goland code to have a helper that returns a golang native enum/iota.

For reference, here is a screenshot from the convo with chatGPT:

![Screenshot 2024-12-02 at 7 33 02 PM](https://github.com/user-attachments/assets/dbd17b60-e32b-4fe9-bdc4-4abe84900069)

Reply:
> I believe it can be simplified to this:
> 
> ```proto
> message Auth {
>   oneof auth_type_details {
>     // Indicates that the GatewayEndpoint does not require authorization.
>     NoAuth no_auth = 1;
>     // Indicates that the GatewayEndpoint uses a static API key for authorization.
>     StaticAPIKey static_api_key = 2;
>     // Indicates that the GatewayEndpoint uses an auth provider issued JWT for authorization.
>     JWT jwt = 3;
>   }
> }
> ```

This looks great and covers all the main review comments.